node/coinstacks/common/api/src/evm/gasOracle.ts

```diff
@@ -0,0 +1,162 @@
+import { ethers } from 'ethers'
+import { Logger } from '@shapeshiftoss/logger'
+import { Fees } from './models'
+import { NewBlock } from '@shapeshiftoss/blockbook'
+
+export interface GasOracleArgs {
+  logger: Logger
+  provider: ethers.providers.JsonRpcProvider
+}
+
+interface TxFees {
+  gasPrices: Array<number>
+  maxPriorityFees: Array<number>
+}
+
+interface BlockFees extends TxFees {
+  baseFeePerGas?: number
+}
+
+type EstimatedFees = Record<string, Fees>
+
+export class GasOracle {
+  private readonly totalBlocks = 20
+  private readonly logger: Logger
+  private readonly provider: ethers.providers.JsonRpcProvider
+
+  private feesByBlock = new Map<string, BlockFees>()
+
+  constructor(args: GasOracleArgs) {
+    this.logger = args.logger.child({ namespace: ['gasOracle'] })
+    this.provider = args.provider
+  }
+
+  async start() {
+    const height = await this.provider.getBlockNumber()
+    const length = this.totalBlocks - 1
+    const startingBlock = height - length
+    const blocks = Array.from<number, string | number>({ length }, (_, index) => index + 1 + startingBlock)
+    blocks.push('pending')
+    await Promise.all(blocks.map(async (blockNumber) => this.update(blockNumber)))
+  }
+
+  getBaseFeePerGas(): string | undefined {
+    return this.feesByBlock.get('pending')?.baseFeePerGas?.toString()
+  }
+
+  // estimate fees based on the current oracle state
+  async estimateFees(percentiles: Array<number>): Promise<EstimatedFees> {
+    const { baseFeePerGas, number } = (await this.provider.send('eth_getBlockByNumber', ['pending', false])) as {
+      number: string
+      baseFeePerGas?: string
+    }
+
+    const latestBlock = Number(number) - 1
+
+    await this.validate(latestBlock)
+
+    if (this.feesByBlock.size !== this.totalBlocks) {
+      this.logger.error(`current blocks: ${this.feesByBlock.size}, expected blocks: ${this.totalBlocks}`)
+    }
+
+    const latestBlockBuffer = latestBlock - 1
+    if (!this.feesByBlock.has(latestBlockBuffer.toString())) {
+      this.logger.error(`oracle is out of sync...`)
+    }
+
+    const estimatedFees = percentiles.reduce<EstimatedFees>((prev, percentile) => {
+      const { gasPrice, maxPriorityFee } = this.averageAtPercentile(percentile)
+      prev[percentile.toString()] = {
+        gasPrice: gasPrice.toFixed(0),
+        maxFeePerGas: (maxPriorityFee + Number(baseFeePerGas)).toFixed(0),
+        maxPriorityFeePerGas: maxPriorityFee.toFixed(0),
+      }
+      return prev
+    }, {})
```

Comment:
Nitpick: can we use a more meaningful name that `prev` in `reduce` blocks? Reduce always fucks with my brain a bit, something like `estimations[percentile.toString()]` is a lot more meaningful out of the box.